### ISSUE TITLE
Cache.crop does not roll back linear-attention recurrent states — block/speculative verification on GDN hybrids (Qwen3-Next / Qwen3.5) is not lossless with DynamicCache

### ISSUE THREAD

**Issue by 0xWhiteMage** [huggingface/transformers#49036]
## System info
- transformers 5.16.1 (also read 5.15.1 source), torch 2.11, CUDA (GB10 / DGX Spark); model type `qwen3_5` (hybrid: 48 Gated-DeltaNet layers `LinearAttentionLayer` + 16 attention layers `DynamicLayer` in `DynamicCache(config=...)`)

## Description
`Cache.crop(tokens_to_remove)` documents: "For layers that do not need to keep all the past states in memory, such as sliding window layers or linear attention layers, this will also restrict the size of the cached states back to their minimal working size", and `activate_past_recording()` says the cache "will wait for a call to `crop` … in order to be able to retrieve previous full states".

For `LinearAttentionCacheLayerMixin` this is only true for **conv states**. `crop` slices `conv_states` back by `tokens_to_remove`, but `recurrent_states` is never touched: the GDN layer (`modeling_qwen3_5.py`, `torch_chunk_gated_delta_rule` / `torch_recurrent_gated_delta_rule` → `cache_params.update_recurrent_state(last_recurrent_state, layer_idx)`) writes the state **after the whole forward** back into the cache, and there is no recorded pre-forward state to return to.

Consequence for any draft-and-verify loop that runs the target over a block of candidate tokens and then crops the rejected suffix (speculative decoding, lookahead, DFlash-style block verification): after every partially accepted block the recurrent state contains the rejected tokens. The target's later logits then depend on tokens that were never emitted, so greedy output depends on the drafter — the loop is not lossless.

## Measurement (Qwen3.5-class 27B hybrid, bf16)
Teacher-forced on one fixed token sequence with two different draft proposal patterns (i.e. different rejected suffixes), same model, same committed tokens, block size 16:
- rejection pattern A: target argmax equals the reference at **384/384** positions (the reference text was produced with this pattern);
- rejection pattern B: target argmax differs at **41/384, 12/384, 38/384, 15/384** positions on four prompts (3–11 %), with top-1 − top-2 logit margins of **0.6 – 2.4** at the first differing positions — a real distribution shift, not bf16 rounding (rounding-level flips have margins 0 – 0.25 here).
- HF `generate` (M=1, recurrent kernel) vs the block loop disagree in the same way; restoring the recurrent states before re-running only the accepted rows removes the effect.

## Minimal reproduction sketch
```python
cache = DynamicCache(config=model.config); cache.activate_past_recording()
model(prefix_ids, past_key_values=cache, use_cache=True)                                  # clean prefix
clean = copy.deepcopy(cache)
model(t1, past_key_values=clean, use_cache=True)                                          # reference: prefix, t1, t2
ref = model(t2, past_key_values=clean, use_cache=True).logits[0, -1]
blk = torch.cat([t1, wrong_ids[:, :k]], dim=1)                                            # "verify" t1 + k rejected drafts
model(blk, past_key_values=cache, use_cache=True); cache.crop(-k)                         # roll the k drafts back
new = model(t2, past_key_values=cache, use_cache=True).logits[0, -1]                      # same tokens as ref: prefix, t1, t2
# -> ref and new differ (argmax flips at moderate margins): cache.layers[i].recurrent_states still include wrong_ids;
#    with an attention-only model the two are identical.
```

## Expected behaviour / suggestions
- When `record_past` is active, record the recurrent state before each `update` (or per token) so `crop` can restore the state at the crop point (memory: one state tensor per recorded step per layer — fine for block sizes ≤ 64); or
- at minimum, make `LinearAttentionCacheLayerMixin.crop` raise (or warn) when `recurrent_states` is initialised, and fix the `Cache.crop` docstring so linear-attention layers are not described as rolled back.

Happy to open a PR for the record-and-restore variant if that direction is acceptable.

**Comment by Rocketknight1** [comment]
This is written in a very 'agent' way that's difficult for humans to read. Can we get a clearer reproducer than that sketch before we decide how to triage this?

**Comment by vasqu** [comment]
Kind of known already and is very core so would rather wait on @Cyrilvallez as he made the counterpart in conv1d related models already. It's needed for MTP models like qwen 3.5+

**Comment by 0xWhiteMage** [comment]
Fair enough, sorry about that — here is a self-contained script instead of the sketch. Tiny random-init `Qwen3_5ForCausalLM`, no downloads, runs on CPU in a few seconds:

<details><summary>repro_cache_crop_gdn.py</summary>

```python
"""Reproducer for huggingface/transformers#49036: Cache.crop() does not roll back linear-attention recurrent state.

Tiny random-init Qwen3.5 text model, CPU, float32, no downloads. The same committed tokens are fed two ways:

    ref : prefix -> t1 -> t2                                      (one token at a time)
    crop: prefix -> [t1, junk_1..junk_k] -> cache.crop(-k) -> t2  (junk tokens verified, then rolled back)

In exact arithmetic the logits at t2 must be identical (t2 sees prefix + t1 in both cases). With attention-only
layers they are. With Gated-DeltaNet ("linear_attention") layers they are not: crop() trims conv_states but
recurrent_states stays as written after the junk tokens. Rolling back to the prefix, restoring the recurrent
states saved there and replaying t1 (what a speculative loop would have to do by hand) makes them identical again.

Usage: python repro_cache_crop_gdn.py            (transformers >= 5.x, torch)
"""
import copy

import torch
from transformers import DynamicCache, Qwen3_5ForCausalLM, Qwen3_5TextConfig

V, K = 128, 12  # vocab, number of junk ("rejected draft") tokens
DEV = "cuda" if torch.cuda.is_available() else "cpu"


def run(layer_types, seed=0):
    torch.manual_seed(seed)
    cfg = Qwen3_5TextConfig(
        vocab_size=V, hidden_size=64, intermediate_size=128, num_hidden_layers=len(layer_types),
        num_attention_heads=4, num_key_value_heads=2, head_dim=16,
        linear_num_key_heads=2, linear_num_value_heads=4, linear_key_head_dim=16, linear_value_head_dim=16,
        linear_conv_kernel_dim=4, layer_types=layer_types,
    )
    model = Qwen3_5ForCausalLM(cfg).eval().to(DEV)
    prefix, t1, t2, junk = (torch.randint(0, V, (1, n), device=DEV) for n in (8, 1, 1, K))

    with torch.no_grad():
        cache = DynamicCache(config=cfg)
        cache.activate_past_recording()
        model(prefix, past_key_values=cache, use_cache=True)
        gdn_layers = [l for l in cache.layers if isinstance(getattr(l, "recurrent_states", None), dict)]
        saved = [{k: v.clone() for k, v in l.recurrent_states.items()} for l in gdn_layers]

        # reference: one token at a time
        ref_cache = copy.deepcopy(cache)
        model(t1, past_key_values=ref_cache, use_cache=True)
        ref = model(t2, past_key_values=ref_cache, use_cache=True).logits[0, -1]

        # verify t1 + K junk tokens in one forward, then roll the junk back with crop()
        model(torch.cat([t1, junk], dim=1), past_key_values=cache, use_cache=True)
        cache.crop(-K)
        after_crop = model(t2, past_key_values=cache, use_cache=True).logits[0, -1]

        # manual fix: roll back to the prefix, restore the recurrent states saved there, replay t1
        cache.crop(-2)  # t2 and t1
        for l, s in zip(gdn_layers, saved):
            l.recurrent_states.update({k: v.clone() for k, v in s.items()})
        model(t1, past_key_values=cache, use_cache=True)
        restored = model(t2, past_key_values=cache, use_cache=True).logits[0, -1]

    diff = lambda a, b: (a - b).abs().max().item()
    return diff(ref, after_crop), diff(ref, restored), (ref.argmax() == after_crop.argmax()).item()


if __name__ == "__main__":
    import transformers
    print(f"transformers {transformers.__version__}  torch {torch.__version__}  device {DEV}")
    for name, lt in (("attention-only  ", ["full_attention"] * 4),
                     ("GDN hybrid (3+1)", ["linear_attention"] * 3 + ["full_attention"])):
        worst = (0.0, 0.0, 0)
        for seed in range(20):
            d_crop, d_fix, same_argmax = run(lt, seed)
            worst = (max(worst[0], d_crop), max(worst[1], d_fix), worst[2] + (not same_argmax))
        print(f"{name}: max |logits_ref - logits_after_crop| = {worst[0]:.3e}   "
              f"after manual recurrent-state restore = {worst[1]:.3e}   argmax flipped in {worst[2]}/20 seeds")
```

</details>

Output on transformers 5.16.1 / torch 2.13 (CPU, pure-torch fallbacks; the GPU/fla path gives the same picture):

```
attention-only  : max |logits_ref - logits_after_crop| = 1.490e-07   after manual recurrent-state restore = 0.000e+00   argmax flipped in 0/20 seeds
GDN hybrid (3+1): max |logits_ref - logits_after_crop| = 2.994e-02   after manual recurrent-state restore = 0.000e+00   argmax flipped in 1/20 seeds
```

So: same committed tokens, but after `crop(-k)` the GDN layers' `recurrent_states` still contain the k cropped tokens, and the next logits differ by ~1e-2 (vs 1e-7 float noise for attention-only layers). Saving the recurrent states at the crop point and putting them back makes the two paths identical again, which is what a speculative/verify loop currently has to do by hand. On a real 27B hybrid we saw greedy argmax flips at top-2 margins of 0.6–2.4 logits from this.

@vasqu understood that it is known and core — happy to wait for @Cyrilvallez. If a record-and-restore in `LinearAttentionCacheLayerMixin` (snapshot per `update` while `record_past` is on, restore in `crop`) is the intended direction, I can put up a PR; otherwise a `crop` that raises for initialised recurrent states plus a docstring fix would at least stop loops from silently going wrong.
